### PR TITLE
Use IntersectionObserver for infinite scroll

### LINKED ISSUE #4537: Improve scroll performance
As per discussion at https://github.com/thepracticaldev/dev.to/pull/4279 im opening this for the future performance optimization in this area.

The problem of accessing dom properties and layout trashing because of that is pretty well described here: https://gist.github.com/paulirish/5d52fb081b3570c81e3a

And as you can see in the PR there is use of some of those properties.

Im not sure how the system works, but if it is run on scroll event (especially if not debounced) then it can become a big issue very quickly, or become an issue when you add/modify your js to do more. Additionally it becomes a little bit harder to find, as in the future you will be editing some other part of the system and suddenly it will show some signs of jank on some devices, and creep upwards (on the device power scale).

Because the PR uses those properties to detect if element is in a certain position, i suggested using a newer API called [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) which doesnt do such things.

There is excellent introduction to this API by Pawel Grzybek: https://pawelgrzybek.com/the-intersection-observer-api-explained/

### PR DESCRIPTION
**Pull request by jack-lewin**
:wave: first-time contributor here!

Resolves #4537.

I can't see any front-end tests which cover this functionality, and I'm not sure where best to add them (if at all). What do you think? I've done some testing locally and would be happy for this to be merged as-is.

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR updates the infinite scroll functionalitiy to use an IntersectionObserver, instead of an interval which was used previously. It also removes some unused variables which accessed scroll-related DOM properties.

Both of these changes should improve scroll performance.

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

n/a - no UI changes

## Added tests?

- [ ] yes
- [ ] no, because they aren't needed
- [x] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/assets/javascripts/initializers/initScrolling.js
Comment (jack-lewin) [inline app/assets/javascripts/initializers/initScrolling.js:420]: I'm not sure what the context was around this `3700px` value - it's been around since the initial commit.

I've added a 500px margin to the `IntersectionObserver` so that new content loads when the user scrolls to within 500px of the page footer.

**Comment by nickytonline** [comment]
@jack-lewin , thanks for the contribution! I’ll be taking a look at this today.

**Comment by nickytonline** [comment]
@jack-lewin, there are some conflicts that need to be resolved.

**Comment by awasum** [comment]
> /rebase

Hi @nickytonline ..This seem to work for you eventhough it was not successful cus Github found the PR not rebaseable here: https://github.com/thepracticaldev/dev.to/runs/830925719?check_suite_focus=true

cc @rhymes

**Comment by nickytonline** [comment]
> > /rebase
> 
> Hi @nickytonline ..This seem to work for you eventhough it was not successful cus Github found the PR not rebaseable here: [thepracticaldev/dev.to/runs/830925719?check_suite_focus=true](https://github.com/thepracticaldev/dev.to/runs/830925719?check_suite_focus=true)
> 
> cc @rhymes

It makes sense it didn't work since there are conflicts.

File: app/assets/javascripts/initializers/initScrolling.js
Comment (nickytonline) [reply app/assets/javascripts/initializers/initScrolling.js:420]: We probably need to adjust this amount of pixels depending on the device. This fine for desktop, but on smaller mobile devices it might be called too frequently. @ben, any idea what the `3700` is supposed to represent?

**Comment by nickytonline** [comment]
Things work locally and this looks good @jack-lewin. Just waiting to hear back from Ben.

**Comment by jack-lewin** [comment]
Thanks for taking a look, @nickytonline!

I've just rebased and fixed the conflict, although I now have some failing tests so will look into that

**Review by rhymes — APPROVED** [review]

**Comment by rhymes** [comment]
@jack-lewin let us know if you need assistance for the failing specs, thank you!

**Comment by nickytonline** [comment]
@Ridhwana can you take a peek at this? I'm a little swamped with frontend PR reviews at the moment.

**Comment by Ridhwana** [comment]
Hi team, I've put this on my radar and will pull the PR down and look through it during the course of today/tomorrow

**Comment by Ridhwana** [comment]
@jack-lewin thank you for this PR 😊 I've taken a look at the PR and tested it, it seems to behave well. I'll approve once the tests are fixed. 

Just wanted to check if you are able to fix the tests or do you need some help :) ?

**Comment by cmgorton** [comment]
@nickytonline @Ridhwana where are we at on this PR. Are there failing tests blocking it?

**Comment by Ridhwana** [comment]
> @nickytonline @Ridhwana where are we at on this PR. Are there failing tests blocking it?

@cmgorton the code looks good, but I was waiting on a response from @jack-lewin to see if he needed help with the test :)

**Comment by msarit** [comment]
Hey @jack-lewin! 👋🏾 
Would you mind looking at the feedback on your PR? It appears the Forem team is willing to help you with tests. 😄 
Let us know if you have any questions

**Comment by mstruve** [comment]
Hey @jack-lewin!, given the lack of activity on this PR I am going to go ahead and close it. Thank you so much for contributing and if you have the time to fix the tests and resolve the conflicts in the future we are more than happy to help you push it through!

**Pull request closed without merging**
